turbopack/crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -401,7 +370,72 @@ impl FileSystem for DiskFileSystem {
             .collect::<Result<_>>()
             .with_context(|| format!("reading directory item in {}", full_path.display()))?;
 
-        Ok(DirectoryContent::new(entries))
+        Ok(InternalDirectoryContent::new(entries))
+    }
+}
+
+impl Debug for DiskFileSystem {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "name: {}, root: {}", self.name, self.root)
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl FileSystem for DiskFileSystem {
+    #[turbo_tasks::function(fs)]
+    async fn read(&self, fs_path: Vc<FileSystemPath>) -> Result<Vc<FileContent>> {
+        let full_path = self.to_sys_path(fs_path).await?;
+        self.register_invalidator(&full_path)?;
+
+        let _lock = self.lock_path(&full_path).await;
+        let content = match retry_future(|| File::from_path(full_path.clone()))
+            .instrument(tracing::info_span!(
+                "read file",
+                path = display(full_path.display())
```

Comment:
Optional nitpick, as I see you're just moving this code around.

```suggestion
                path = %full_path.display()
```

![Screenshot 2024-09-04 at 10.40.12 AM.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/HAZVitxRNnZz8QMiPn4a/08a99d52-3123-423f-9945-c5d9909989e5.png)